File: 21553 - Sai Nandan D/Week 6/23_Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Approach 1: Priority Queue
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        ListNode.__lt__ = lambda self, other: self.val < other.val
        
        result_head = result_tail = ListNode()
        heap = []
        for list_head in lists:
            if list_head:
                heapq.heappush(heap, list_head)
                
        while heap:
            node = heapq.heappop(heap)
            result_tail.next = node
            result_tail = result_tail.next
            if node.next:
                heapq.heappush(heap, node.next)
                
        return result_head.next
        
 
 # Approach 2: Divide and Conquer
 
 # 2.1: Iterative merging
 class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        def merge2lists(l1, l2):
            head = result = ListNode()
            while l1 and l2:
                if l1.val <= l2.val:
                    result.next = l1
                    l1 = l1.next
                else:
                    result.next = l2
                    l2 = l2.next                    
                result = result.next
            result.next = l1 or l2
            return head.next
        
        n = len(lists)
        step = 1        
        while step < n:
            for i in range(0, n-step, step*2):
                lists[i] = merge2lists(lists[i], lists[i+step])
            step *= 2
            
        return lists[0] if n > 0 else None
        
 # 2.2: Recursive merging
 class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        def merge2lists(l1, l2):
            if l1 and l2:
                if l1.val > l2.val:
                    l1, l2 = l2, l1
                l1.next = merge2lists(l1.next, l2)
            return l1 or l2
            
        
        n = len(lists)
        step = 1        
        while step < n:
            for i in range(0, n-step, step*2):
                lists[i] = merge2lists(lists[i], lists[i+step])
            step *= 2
            
        return lists[0] if n > 0 else None

File: 21558 - Nithin/week 6/23-merge-k-sorted-lists/23-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int k = lists.size();
        if(k == 0)
            return NULL;
        priority_queue<int, vector<int>, greater<int>> pq_merged;
        for(int i = 0;i < k;i++){
            ListNode* node = lists[i];
            while(node){
                pq_merged.push(node->val);
                node = node->next;
            }
        }
        if(pq_merged.size() == 0)
            return NULL;
        
        ListNode* merged1 = new ListNode();
        ListNode* merged = merged1;
        while(pq_merged.size()){
            merged->next = new ListNode(pq_merged.top());
            pq_merged.pop();
            merged = merged->next;
        }
        merged->next = NULL;
        merged1 = merged1->next;
        return merged1;
        
    }
};